src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatProvider.cs

```diff
@@ -0,0 +1,354 @@
+            {
+                try
+                {
+                    return this.AddHealthPropertyInternal(this.extendPayloadItems, name, value, isHealthy);
+                }
+                catch (Exception e)
+                {
+                    CoreEventSource.Log.LogError(
+                        string.Format(CultureInfo.CurrentCulture,
+                        "Failed to set a health heartbeat property named '{0}'. Exception: {1}",
+                            name,
+                            e.ToInvariantString()));
+                }
+            }
+
+            return false;
+        }
+
+        public bool SetHealthProperty(string name, string value = null, bool? isHealthy = null)
+        {
+            if (!string.IsNullOrEmpty(name))
+            {
+                try
+                {
+                    return this.SetHealthPropertyInternal(this.extendPayloadItems, name, value, isHealthy);
+                }
+                catch (Exception e)
+                {
+                    CoreEventSource.Log.LogError(
+                        string.Format(CultureInfo.CurrentCulture,
+                        "Failed to set a health heartbeat property named '{0}'. Exception: {1}",
+                            name,
+                            e.ToInvariantString()));
+                }
+            }
+
+            return false;
+        }
+
+        public bool RemoveHealthProperty(string payloadItemName)
+        {
+            if (!string.IsNullOrEmpty(payloadItemName))
+            {
+                try
+                {
+                    return this.extendPayloadItems.TryRemove(payloadItemName, out HeartbeatPropertyPayload removedItem);
+                }
+                catch (Exception e)
+                {
+                    CoreEventSource.Log.LogError(
+                        string.Format(CultureInfo.CurrentCulture,
+                        "Failed to remove a health heartbeat property named '{0}'. Exception: {1}",
+                            payloadItemName,
+                            e.ToInvariantString()));
+                }
+            }
+
+            return false;
+        }
+
+        #region IDisposable Support
+
+        public void Dispose()
+        {
+            this.Dispose(true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!this.disposedValue)
+            {
+                if (disposing)
+                {
+                    if (this.HeartbeatTimer != null)
+                    {
+                        this.isEnabled = false;
+                        this.HeartbeatTimer.Dispose();
+                    }
+                }
+
+                this.disposedValue = true;
+            }
+        }
+
+        #endregion
+
+        protected virtual void Send()
+        {
+            this.InternalSendHealthHeartbeat(this.GatherData());
+        }
+
+        protected virtual MetricTelemetry GatherData()
```

Comment:
these methods are just helpers and can be private right?

Reply:
They were there as something for me to implement in a 'mock' test class that extended the provider. Instead, I've made this function 'internal' and documented the reason for it. I believe this will be a better approach to unit testing.